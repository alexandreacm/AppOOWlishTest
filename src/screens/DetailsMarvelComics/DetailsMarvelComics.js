import React, { useEffect, useMemo } from 'react';
import styled from 'styled-components/native';
import { ActivityIndicator } from 'react-native';

import {
  useRoute,
  useIsFocused,
  useNavigation
} from '@react-navigation/native';
import { useSelector, useDispatch } from 'react-redux';

import colors from '@/theme/colors';
import Header from '@/components/Header';
import CardDetail from '@/components/CardDetail';

import { LOADING_DETAILS_CHARACTERS } from '@/store/slices/marvelComicsSlice';

const DetailsMarvelComics = () => {
  const isFocused = useIsFocused();
  const { goBack } = useNavigation();
  const dispatch = useDispatch();

  const { isLoading, detailsCharactersData, errorMessage } = useSelector(
    ({ comics }) => comics
  );

  const {
    params: { id }
  } = useRoute();

  const handleGoBack = () => goBack();

  const contentDetailComic = useMemo(() => {
    if (isLoading)
      return <ActivityIndicator color={colors.SECUNDARY} size='large' />;

    return detailsCharactersData.map(detailCharactersData => (
      <CardDetail
        key={detailCharactersData?.id}
        image={{
          uri: `${detailCharactersData?.thumbnail.path}.${detailCharactersData?.thumbnail.extension}`
        }}
        title={detailCharactersData?.title}
        issueNumber={detailCharactersData?.issueNumber}
        price={detailCharactersData?.prices[0].price}
        description={
          detailCharactersData?.description
            ? detailCharactersData?.description
            : 'Sem descrição'
        }
      />
    ));
  }, [isLoading, detailsCharactersData]);

  useEffect(() => {
    dispatch(LOADING_DETAILS_CHARACTERS({ id }));
  }, [dispatch, id]);

  return (
    <>
      <Header
        showBackButton
        leftIcon
        onBackPress={handleGoBack}
        backgroundColor={colors.PRIMARY}
        backButtonColor={colors.COLOR_WHITE}
        isFocuse={isFocused}
      />
      {errorMessage && (
        <StyledViewLabel>
          <StyledLabel>{errorMessage}</StyledLabel>
        </StyledViewLabel>
      )}
      <StyledScrollView showsVerticalScrollIndicator={false}>
        <StyledContainerDefault>{contentDetailComic}</StyledContainerDefault>
      </StyledScrollView>
    </>
  );
};

const StyledContainerDefault = styled.View`
  flex: 1;
  background-color: ${colors.COLOR_WHITE};
  margin-left: 6px;
  margin-right: 6px;
  margin-top: 10px;
`;

const StyledLabel = styled.Text`
  width: 100%;
  font-weight: 400;
  font-size: 14px;
  margin-bottom: 8px;
  color: ${colors.PRIMARY};
`;

const StyledViewLabel = styled.View`
  width: 100%;
  padding: 10px;
  align-items: center;
  background-color: ${colors.DANGER};
`;

const StyledScrollView = styled.ScrollView`
  flex: 1;
  background: ${colors.COLOR_WHITE};
`;
export default DetailsMarvelComics;
