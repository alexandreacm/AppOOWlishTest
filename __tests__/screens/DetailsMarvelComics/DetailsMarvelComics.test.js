import 'react-native';
import React from 'react';
import {
  matchSnapshotWithTheme,
  renderWithTheme
} from '@/helpers/functions/testing';

import DetailsMarvelComics from '@/screens/DetailsMarvelComics';

describe('tests in DetailsMarvelComics Screen', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(<DetailsMarvelComics />);
  });

  test('renders correctly', () => {
    renderWithTheme(<DetailsMarvelComics />);
  });
});
