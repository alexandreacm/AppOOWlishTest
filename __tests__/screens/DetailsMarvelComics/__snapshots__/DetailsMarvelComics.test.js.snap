// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in DetailsMarvelComics Screen matches snapshot 1`] = `
Array [
  <RCTSafeAreaView
    backgroundColor="#FF911E"
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#FF911E",
        },
      ]
    }
  />,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "height": 20,
        },
        Object {
          "backgroundColor": "#FF911E",
        },
      ]
    }
  />,
  <View
    backgroundColor="#FF911E"
    slim={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FF911E",
          "flexDirection": "row",
          "height": 56,
          "justifyContent": "space-between",
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 0,
          "width": "100%",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#FFF",
              "fontSize": 26,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î—„
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 37,
            "justifyContent": "center",
            "marginRight": 27,
            "opacity": 1,
            "width": 37,
          }
        }
        testID="left-icon"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/icMessage.png",
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 37,
            "justifyContent": "center",
            "marginRight": 27,
            "opacity": 1,
            "width": 37,
          }
        }
        testID="right-icon"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/icBell.png",
            }
          }
        />
      </View>
    </View>
  </View>,
  <RCTScrollView
    showsVerticalScrollIndicator={false}
    style={
      Array [
        Object {
          "backgroundColor": "#FFF",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFF",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "marginLeft": 6,
              "marginRight": 6,
              "marginTop": 10,
            },
          ]
        }
      />
    </View>
  </RCTScrollView>,
]
`;
